prereise/gather/flexibilitydata/geo/geo_data.py

```diff
@@ -0,0 +1,253 @@
+import pickle as pkl
+import pandas as pd
+import numpy as np
+import scips as sp
+import geopy
+import requets
+import time
+from geopy.extra.rate_limiter import RateLimiter
+
+"""
+    eiaid_to_zip()
+
+Find the service region (list of ZIP codes) for every LSE identified by their EIA ID
+Create and save a dictionary with EIA ID as keys for list of zip codes 
+"""
+def eiaid_to_zip():
+    iou_path = "../raw/iou_zipcodes_2019.csv"
+    niou_path = "../raw/non_iou_zipcodes_2019.csv"
+    dictpath = "../cache/eiaid2zip.pkl"
+
+    iou_df = pd.read_csv(iou_path)
+    niou_df = pd.read_csv(niou_path)
+
+    all_ids = list(set(iou_df['eiaid'].to_list() + niou_df['eiaid'].to_list()))
+
+    id2zip = {}
+
+    for i in all_ids:
+        iouzips = iou_df.loc[iou_df['eiaid']==i]['zip'].to_list()
+        niouzips = niou_df.loc[niou_df['eiaid']==i]['zip'].to_list()
+        allzips = list(set(iouzips + niouzips))
+        allzips.sort()
+        id2zip[i] = allzips
+
+    with open(dictpath, 'wb') as fh:
+        pickle.dump(id2zip, fh)
+
+"""
+    get_bus_pos(case_path)
+
+Read a .mat case and extract the lat/lon coordinate of all buses
+return a list of (busID, lat, lon)
+"""
+def get_bus_pos(case_path):
+    case = spio.loadmat(case_path)
+    mpc = case['mpc']
+
+    bus = mpc['bus'][0][0]
+    bus_num = bus.shape[0]
+    bus2sub = mpc['bus2sub'][0][0]
+    sub = mpc['sub'][0][0]
+    subid = mpc['subid'][0][0]
+
+    bus_pos = np.zeros((bus_num, 3))
+
+    # [id, lat, long]
+    subid_list = subid[:, 0].tolist()
+    for i in range(bus_num):
+            
+        bus_sub_idx = subid_list.index(bus2sub[i, 0])
+        bus_pos[i, 0] = bus[i, 0]
+        bus_pos[i, 1] = sub[bus_sub_idx, 2]
+        bus_pos[i, 2] = sub[bus_sub_idx, 3]
+
+    return bus_pos
+
+
+"""
+    get_bus_fips(case_path)
+
+Try to get FIPS of each bus in a case mat using FCC AREA API
+Can take hours to run, save to cache file for future use
+"""
+def get_bus_fips(case_path):
+
+    bus_pos = get_bus_pos(case_path)
+    bus_fips_dict = {'busid': bus_pos[:, 0],
+                    'latitude': bus_pos[:, 1],
+                    'longitude': bus_pos[:, 2],
+                    'fips': [0]*bus_num}
+        
+        
+    url = "https://geo.fcc.gov/api/census/area"
+    start_idx = 0
+
+    for i in range(start_idx, bus_num):
```

Comment:
this is used to resume a session where the server failed in the middle of a query process. This will need to be addressed in the main file by calling this function again in case the server fails, and start idx will be needed there.

Reply:
Okay, so should `start_idx` be taken as an input by `get_bus_fips`?

Reply:
yes, I have changed that in the upcoming commit. 
As I mentioned in another reply these are from a big jupyter notebook I used to develop and test the functionalities. They work but there could be a lot of messes not cleaned in this version.